sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,383 @@
+    length = 4;
+    mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET,
+                        address);
+  } else {
+    length = 3;
+    address >>= 8;
+    mmio_region_write32(spi_host->base_addr, SPI_HOST_TXDATA_REG_OFFSET,
+                        address);
+  }
+  write_command_reg(spi_host, length, segment->address.width,
+                    kDifSpiHostDirectionTx, last_segment);
+}
+
+static void issue_dummy(const dif_spi_host_t *spi_host,
+                        dif_spi_host_segment_t *segment, bool last_segment) {
+  write_command_reg(spi_host, segment->dummy.length, segment->dummy.width,
+                    kDifSpiHostDirectionDummy, last_segment);
+}
+
+static void issue_data_phase(const dif_spi_host_t *spi_host,
+                             dif_spi_host_segment_t *segment,
+                             bool last_segment) {
+  size_t length;
+  dif_spi_host_width_t width;
+  dif_spi_host_direction_t direction;
+
+  switch (segment->type) {
+    case kDifSpiHostSegmentTypeTx:
+      width = segment->tx.width;
+      length = segment->tx.length;
+      direction = kDifSpiHostDirectionTx;
+      spi_host_fifo_write_alias(spi_host, segment->tx.buf, segment->tx.length);
+      break;
+    case kDifSpiHostSegmentTypeBidirectional:
+      width = segment->bidir.width;
+      length = segment->bidir.length;
+      direction = kDifSpiHostDirectionBidirectional;
+      spi_host_fifo_write_alias(spi_host, segment->bidir.txbuf,
+                                segment->bidir.length);
+      break;
+    case kDifSpiHostSegmentTypeRx:
+      width = segment->rx.width;
+      length = segment->rx.length;
+      direction = kDifSpiHostDirectionRx;
+      break;
+    default:
+      // Programming error.  We should never get here.
+      // LOG_FATAL("issue_data_phase: bad segment type %d", segment->type);
+      abort();
+  }
+  write_command_reg(spi_host, length, width, direction, last_segment);
+}
+
+dif_result_t dif_spi_host_transaction(const dif_spi_host_t *spi_host,
+                                      uint32_t csid,
+                                      dif_spi_host_segment_t *segments,
+                                      size_t length) {
+  // Write to chip select ID.
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CSID_REG_OFFSET, csid);
+
+  // For each segment, write the segment information to the
+  // COMMAND register and transmit FIFO.
+  for (size_t i = 0; i < length; ++i) {
+    bool last_segment = i == length - 1;
+    wait_ready(spi_host);
+    dif_spi_host_segment_t *segment = &segments[i];
+    switch (segment->type) {
+      case kDifSpiHostSegmentTypeOpcode:
+        issue_opcode(spi_host, segment, last_segment);
+        break;
+      case kDifSpiHostSegmentTypeAddress:
+        issue_address(spi_host, segment, last_segment);
+        break;
+      case kDifSpiHostSegmentTypeDummy:
+        issue_dummy(spi_host, segment, last_segment);
+        break;
+      case kDifSpiHostSegmentTypeTx:
```

Comment:
I believe that `-Wfallthrough` is completely ok with code of the form
```c
case kFoo:
case kBar:
case kBaz: AyyLmao(); break;
```

Reply:
Yeah, that works.  I'd originally added the FALLTHROUGH_INTENDED macro for the alert and shutdown handler, but those fallthrough cases weren't trivial like these.